src/persist-txn/src/txn_cache.rs

```diff
@@ -1046,127 +1046,118 @@ mod tests {
-        // assert_eq!(cache.data_listen_next(&d0, 5), WaitForTxnsProgress);
-
-        // Advance time. Now we can resolve queries at higher times. Nothing got
-        // written at 5, so the physical upper hasn't moved, but we're now
-        // guaranteed that `[5, 6)` is empty.
-        cache.progress_exclusive = 6;
-        assert_eq!(cache.data_listen_next(&d0, 5), EmitLogicalProgress(6));
-
-        // Write a batch. Apply will advance the physical upper, so now we need
-        // to wait for the data upper to advance past the commit ts.
-        cache.push_append(d0, vec![0xA0], 6, 1);
-        cache.progress_exclusive = 7;
-        assert_eq!(cache.data_listen_next(&d0, 6), ReadDataTo(7));
-
-        // An unrelated batch is another empty space guarantee. Ditto retracting
-        // the batch.
-        let other = ShardId::new();
-        cache.push_register(other, 7, 1, 7);
-        cache.push_append(other, vec![0xB0], 7, 1);
-        cache.push_append(d0, vec![0xA0], 7, -1);
-        cache.progress_exclusive = 8;
-        assert_eq!(cache.data_listen_next(&d0, 7), EmitLogicalProgress(8));
-
-        // Unlike a snapshot, the previous answers have changed! This is because
-        // it's silly (inefficient) to walk through the state machine for a big
-        // range of time once you know of a later write that advances the
-        // physical upper.
-        // assert_eq!(cache.data_listen_next(&d0, 3), ReadDataTo(7));
-        // assert_eq!(cache.data_listen_next(&d0, 4), ReadDataTo(7));
-        assert_eq!(cache.data_listen_next(&d0, 5), ReadDataTo(7));
-        assert_eq!(cache.data_listen_next(&d0, 6), ReadDataTo(7));
+        // This attempts to exercise all the various interesting edge cases of
+        // data_snapshot and data_listen_subscribe using the following sequence
+        // of events:
+        //
+        // - Registrations at: [2,8], [12,13]
+        // - Direct writes at: 1, 10
+        // - Writes via txns at: 4, 5, 7
+        let mut c = TxnsCacheState::new(ShardId::new(), 0, None);
+
+        // empty
+        assert_eq!(c.progress_exclusive, 0);
+        assert!(std::panic::catch_unwind(|| c.data_snapshot(d0, 0)).is_err());
+        assert_eq!(c.data_listen_next(&d0, 0), WaitForTxnsProgress);
+
+        // ts 0 (never registered)
+        testcase(&mut c, 0, d0, ds(None, 0, 1), ReadDataTo(1));
+
+        // ts 1 (direct write)
+        testcase(&mut c, 1, d0, ds(None, 1, 2), ReadDataTo(2));
+
+        // ts 2 (register)
+        c.push_register(d0, 2, 1, 2);
+        testcase(&mut c, 2, d0, ds(None, 2, 3), EmitLogicalProgress(3));
+
+        // ts 3 (registered, not written)
+        testcase(&mut c, 3, d0, ds(None, 3, 4), EmitLogicalProgress(4));
+
+        // ts 4 (written via txns)
+        c.push_append(d0, vec![4], 4, 1);
+        testcase(&mut c, 4, d0, ds(Some(4), 4, 5), ReadDataTo(5));
+
+        // ts 5 (written via txns, write at preceding ts)
```

Comment:
why is it interesting that we have to txn writes back-to-back? I forgot why we wanted that :see_no_evil: